Fix BPF Map Pressure for policy maps goes above 100%

tommyp1ckles opened the pull request:
endpoint: fix issue where bpf_map_pressure goes above 100% for policy

Endpoints that are being created/restored would have their
realizedPolicy and desiredPolicy to point at the same object in memory.

Under normal operations, this seems to not cause much issue since
everything that is intended to be in the desired policy state should
also end up in the realized policy state.

However, in the case where the bpf policy map fills up, the realized
policy map would appear to grow above its max size as the desired map
state would still be updated.

In such a case, this was noticeable with the bpf_map_pressure metric for
policy map types going above 100%.

This can be reproduced by setting bpf.policyMapMax=256 and creating a
Pod using fqdn policy and continuously performing DNS lookups from
within the Pod to domains that return many IPs (i.e. such as AWS
domains).
Eventually the policy map should fill up, continuing doing lookups
further will lead to unexpected bpf_map_pressure metrics for policy
maps.

**Note:** Submitting #28185 as a seperate PR to make backporting easier (this change will likely need to be backported, the former will not) as a series of changes to fix the map pressure metric for policy maps.

tommyp1ckles committed:
endpoint: fix issue where bpf_map_pressure goes above 100% for policy

Endpoints that are being created/restored would have their
realizedPolicy and desiredPolicy to point at the same object in memory.

Under normal operations, this seems to not cause much issue since
everything that is intended to be in the desired policy state should
also end up in the realized policy state.

However, in the case where the bpf policy map fills up, the realized
policy map would appear to grow above its max size as the desired map
state would still be updated.

In such a case, this was noticeable with the bpf_map_pressure metric for
policy map types going above 100% as a result of the realizedPolicy
being inserted into via the desiredPolicy reference despite it no longer
reflecting the actual realized bpf map.

This can be reproduced by setting bpf.policyMapMax=256 and creating a
Pod using fqdn policy and continuously performing DNS lookups from
within the Pod to domains that return many IPs (i.e. such as AWS
domains).
Eventually the policy map should fill up, continuing doing lookups
further will lead to unexpected bpf_map_pressure metrics for policy
maps.

This commit makes the following changes:

1. desiredPolicy and realizedPolicy no longer ever point to the same
   endpoint policy.  It is now expected that these are managed
   separately where realized policy is updated following a syncPolicyMap
   call and should reflect actual realized state in the endpoint policy
   map.

2. Desired and realized policy map states are no longer compared by
   pointer value, it is now assumed that realized state should be
   managed when doing syncPolicyMap().

3. When running on drymode, realized policy is automatically copied from
   the desired policy map - under the assumption that syncPolicyMap
   would have succeeded.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>

christarazi commented on pkg/endpoint/endpoint.go:
```diff
@@ -516,7 +519,7 @@ func createEndpoint(owner regeneration.Owner, policyGetter policyRepoGetter, nam
 	ep.aliveCtx = ctx
 
-	ep.realizedPolicy = ep.desiredPolicy
+	ep.realizedPolicy = ep.desiredPolicy.DeepCopy()
```
While I understand this is probably technically the most correct thing to do, a deep-copy is quite expensive, especially in the hot path. Could we explore a cheaper alternative to exceeding 100%? Maybe during the metrics-update, we check whether the current size exceeds the max, and then truncate?

My guess is that historically we have may have avoided deep copies due to the performance implications.

cc @jrajahalme @joestringer for background.

tommyp1ckles replied:
I was a bit surprised to find this, but from what I can tell realizedPolicy and desiredPolicy always point at the same thing.  As a result, realizedPolicy is basically functioning the same as desiredPolicy - perhaps we could just get rid of realized policy?

But I'm missing some context here so I'd like to see what Joe or Jarno think.

aditighag reviewed (commented):
There are issues highlighted by the go linter. Do you want to put this into draft in the meantime?

tommyp1ckles committed:
endpoint: map pressure metric should use desired state.

Per discussion here: https://github.com/cilium/cilium/pull/28184#issuecomment-1768913800

bpf_map_pressure should actually reflect desired state, as it does in
the default implementation in pkg/bpf.

Following previous commit that seperates realized vs desired state, this
changes endpoint specific bpf_map_pressure update implementation to
reflect desired state.

Signed-off-by: Tom Hadlaw <tom.hadlaw@isovalent.com>
